File: ctypes/src/VanDerPol.c
// VanDerPol.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include "../include/ODE.h"

typedef struct 
{
	double mu;
} VanDerPol_p_t;

void VanDerPol_fun(const double t, const double *x, double *f, const void* p)
{
    // Unpack pointer
    VanDerPol_p_t* VanDerPol_p;
	VanDerPol_p = (VanDerPol_p_t*) p;
    // Calculate derivative
    f[0] = x[1];
    f[1] = (VanDerPol_p->mu)*(1-x[0]*x[0])*x[1] - x[0];
}

void VanDerPol(double* x, double* t, double dt, int Ns, double mu)
{
    // Create and declare void pointer
    VanDerPol_p_t VanDerPol_p;
    VanDerPol_p_t* p = &VanDerPol_p;
    // Fill in value
    p->mu = mu;
    // Declare function
    fun_t* fun = VanDerPol_fun;
    // Number of variables
    int nx = 2;
    // Initialize f
    double * f = (double*) malloc(nx*sizeof(double));
    // Simulate
    ExplicitEuler(fun,f,x,nx,t,dt,Ns,p);
    // free f
    free(f);
}

File: python/pylib/VanDerPol.py
import numpy as np
from pylib.ExplicitEuler import ExplicitEuler


def VanDerPol_fun(t,x,p):
    # Calculate slope
    f = np.array([x[1],p*(1-x[0]*x[0])*x[1]-x[0]])
    # Return
    return f


def VanDerPol(x0,t0,dt,Ns,mu):
    # Allocate memory
    x = np.empty((2,Ns+1),dtype=float)
    t = np.empty(Ns+1,dtype=float)
    # Initialize initial conditions
    x[:,0] = x0
    t[0]   = t0
    # Simulate
    ExplicitEuler(VanDerPol_fun,x,t,dt,Ns,mu)
    
    return x,t